feefrac: add support for evaluating at given size

sipa opened the pull request:
The `FeeFrac` type represents a fraction, intended to be used for sats/vbyte or sats/WU. This PR adds functionality to evaluate that feerate for a given size, in order to obtain the fee it corresponds with (rounding down, or rounding up).

The motivation here is being able to do accurate feerate evaluations in cluster mempool block building heuristics (where rounding down is needed), but in principle this makes it possible to use `FeeFrac` as a more accurate replacement for `CFeeRate` (where for feerate estimation rounding up is desirable). Because of this, both rounding modes are implemented.

Unit tests are included for known-correct values, plus a fuzz test that verifies the result using `arith_uint256`.

sipa committed:
feefrac: add helper functions for 96-bit division

These functions are needed to implement FeeFrac evaluation later: given a
FeeFrac{fee, size}, its fee at at_size is (fee * at_size / size).

ismaelsadeeq reviewed (commented):
Concept ACK

I think we should also have  `GetFeePerK` or `GetFeePerS` method in `FeeFrac` similar to `CFeeRate`

tdb3 reviewed (commented):
Concept ACK
Seems like a helpful improvement.
Would like to circle back to take a deeper look.
Out of curiosity, why the adjustment from 0x200000004 to 0x200000000? (nothing immediately popped out to me)

paplorinc commented on src/util/feefrac.h:
```diff
@@ -144,6 +144,43 @@ struct FeeFrac
+            // this division to round down however, while the / operator rounds towards zero. In
+            // case low64 is negative and not a multiple of size, we thus need a correction.
+            int64_t low64_div = low64 / size;
+            low64_div -= (low64 % size) < 0;
```
oh, wow, I had to look at this twice :D

paplorinc commented on src/util/feefrac.h:
```diff
@@ -144,6 +144,43 @@ struct FeeFrac
+            int64_t high64_div = high64 / size;
+            // Evaluate the parenthesized expression above, so the result becomes
+            // low64 / this->size + (high64_div * 2**32)
+            int64_t low64 = ((high64 - high64_div * size) << 32) + low32;
```
nit: while the return value needs a `+`, I think this would work with a ` | low32` as well (which would provide better documentation about them not overlapping)

paplorinc commented on src/test/feefrac_tests.cpp:
```diff
@@ -17,6 +17,27 @@ BOOST_AUTO_TEST_CASE(feefrac_operators)
+
+    FeeFrac neg{-1001, 100};
+    BOOST_CHECK_EQUAL(neg.Evaluate(0), 0);
+    BOOST_CHECK_EQUAL(neg.Evaluate(1), -11);
```
how often would we get negative fee in reality? If rounding towards negative infinity just a theoretical exercise or is it an important usecase?

l0rinc reviewed (changes requested):
I think we could make the code more testable by extracting the 96 bit division like we did for the multiplication

ceffikhan commented:
Concept ACK

sipa committed:
feefrac: rework comments around Mul/MulFallback

glozow reviewed (commented):
Concept ACK 

Also concept ack to changing `CFeeRate` to use a `FeeFrac` internally. Did it locally to see what the quirks are - `CFeeRate` rounds up, is that maybe worth matching?

sipa committed:
arith_uint256: modernize comparison operators

Since C++20, operator!= is implicitly defaulted using operator==, and
operator<, operator<=, operator>, and operator>= are defaulted using
operator<=>, so it suffices to just provide these two.

sipa committed:
feefrac fuzz: use arith_uint256 instead of ad-hoc multiply

Rather than use an ad-hoc reimplementation of wide multiplication inside the
fuzz test, reuse arith_uint256, which already has this. It's larger than what we
need here, but performance isn't a concern in this test, and it does what we need.

sipa committed:
feefrac: add support for evaluating at given size

sipa committed:
feefrac: support both rounding up and down for Evaluate

Co-Authored-By: l0rinc <pap.lorinc@gmail.com>

paplorinc commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -59,23 +48,14 @@ std::strong_ordering MulCompare(int64_t a1, int64_t a2, int64_t b1, int64_t b2)
-    // intentional here.
-    if (a1 < 0) abs_a1 = ~abs_a1 + 1;
-    if (a2 < 0) abs_a2 = ~abs_a2 + 1;
-    if (b1 < 0) abs_b1 = ~abs_b1 + 1;
-    if (b2 < 0) abs_b2 = ~abs_b2 + 1;
+    // Compute absolute values of products.
```
Is it important to distinguish the 0 values from the positive ones when comparing the signs?
If not, maybe we can simplify a bit:
```C++
    // Compute and compare signs.
    int sign_a = (a1 < 0 ? -1 : 1) * (a2 < 0 ? -1 : 1);
    int sign_b = (b1 < 0 ? -1 : 1) * (b2 < 0 ? -1 : 1);
    if (sign_a != sign_b) return sign_a <=> sign_b;

    auto result = Abs256(a1) * Abs256(a2) <=> Abs256(b1) * Abs256(b2);
    return (sign_a >= 0) ? result : 0 <=> result;

paplorinc replied:
In that case consider only the second part of the suggestion:
```C++
    auto result = Abs256(a1) * Abs256(a2) <=> Abs256(b1) * Abs256(b2);
    return (sign_a >= 0) ? result : 0 <=> result;
```
(obviously non-blocker)

paplorinc replied:
Or a local lambda, for the first part to make it cleaner:
```C++
    auto sign = [](int64_t x) { return x == 0 ? 0 : x < 0 ? -1 : 1; };
    int sign_a = sign(a1) * sign(a2);
    int sign_b = sign(b1) * sign(b2);
    if (sign_a != sign_b) return sign_a <=> sign_b;
```

paplorinc commented on src/util/feefrac.h:
```diff
@@ -182,6 +182,27 @@ struct FeeFrac
+    {
+        Assume(size > 0);
+        Assume(at_size >= 0);
+        if (fee >= 0 && fee < 0x200000000) [[likely]] {
```
would it make sense to add non-fuzz `0x1ffffffff`, `0x200000000` and `0x200000001` test cases for this boundary?

l0rinc reviewed (approved):
ACK 1cc0d5475a578a8299c24c008c0080a6006cae28

I left a few recommendations and questions, was really fun reviewing this change - though it took me a lot of time.

If you think our contributions have affected the change, please consider adding us as co-authors.

paplorinc commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -121,3 +104,89 @@ FUZZ_TARGET(feefrac)
+    auto prod_abs = Abs256(mul32) * Abs256(mul64);
+    auto div_abs = Abs256(div);
+    auto quot_abs = (is_negative == round_down) ?
+        (prod_abs + div_abs - 1) / div_abs :
```
I meant that in https://github.com/bitcoin/bitcoin/pull/30535/files#diff-09e6cf871236bf03d32cca9405837d9b7927690b2296a2de17c9be6ea0e75959R215 we're using the same `A + (B - 1) / B` equation for rounding, i.e. the test and production code may be too similar.
I think it's fine since the big fee rounding is the more complicated one.

paplorinc replied:
Indeed, thanks for checking

paplorinc commented on src/util/feefrac.h:
```diff
@@ -144,6 +191,38 @@ struct FeeFrac
+            }
+        } else {
+            // Otherwise, use Mul and Div.
+            return Div(Mul(fee, at_size), size, RoundDown);
```
Does it test `return Div(Mul(fee, at_size), size, RoundDown)` when `fee < 0x200000000`?
I don't think that can happen in reality, I just tested it to make sure it's correct for small values as well.

paplorinc replied:
>  But please, it's one line.

+1 meant that I like this, maybe I should use a 👍  next time :)

paplorinc replied:
Thanks!

paplorinc commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -121,3 +104,89 @@ FUZZ_TARGET(feefrac)
+FUZZ_TARGET(feefrac_mul_div)
+{
+    // Verify the behavior of:
+    // - The combination of FeeFrac::Mul + FeeFrac::Div.
+    // - The combination of FeeFrac::MulFallback + FeeFrac::DivFallback.
```
My mistake, thanks for clarifying

paplorinc commented:
ACK 9590cdd1afc78ff73fdccd2d016075bf40b6b0e2

brunoerg commented:
Concept ACK

brunoerg commented:
I didn't review it in depth but I used this PR to test my "mutation testing tool", which creates mutations based on diff, and the tests are REALLY great! The only two mutants that were not killed seems to be "equivalent mutants", so we can ignore them. See:

------------------------------
#### Bitcoin Core (#30535 - feefrac: add support for evaluating at given size)

* `make && ./src/test/test_bitcoin --run_test=feefrac_tests` 

* Mutation score: **94.4%** (**Excelent**)

--------------------------------------------

#### MUTANTS NOT KILLED

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.22.h
index dbb03bbccc..0930d12846 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.22.h
@@ -201,7 +201,7 @@ struct FeeFrac
     {
         Assume(size > 0);
         Assume(at_size >= 0);
-        if (fee >= 0 && fee < 0x200000000) [[likely]] {
+        if (fee >= 0 && fee <= 0x200000000) [[likely]] {
             // Common case where (this->fee * at_size) is guaranteed to fit in a uint64_t.
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.13.h
index dbb03bbccc..71cb90757e 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.13.h
@@ -201,7 +201,7 @@ struct FeeFrac
     {
         Assume(size > 0);
         Assume(at_size >= 0);
-        if (fee >= 0 && fee < 0x200000000) [[likely]] {
+        if (fee > 0 && fee < 0x200000000) [[likely]] {
             // Common case where (this->fee * at_size) is guaranteed to fit in a uint64_t.
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);

```

--------------------------------------------

<details>
<summary>Killed mutants</summary>
<br>


```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.27.h
index dbb03bbccc..1f9ca50a5e 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.27.h
@@ -69,7 +69,7 @@ struct FeeFrac
         // that the / operator here rounds in the wrong direction (if n_low is not a multiple of
         // size, and is (if round_down) negative, or (if !round_down) positive). If so, make a
         // correction.
-        int64_t quot_low = n_low / d;
+                int64_t quot_low = (n_low / d) + 1;
         int32_t mod_low = n_low % d;
         quot_low += (mod_low > 0) - (mod_low && round_down);
         // Combine and return the result

```
```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.29.h
index dbb03bbccc..96f39cdd74 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.29.h
@@ -92,7 +92,7 @@ struct FeeFrac
     {
         Assume(d > 0);
         // Compute the division.
-        int64_t quot = n / d;
+                int64_t quot = (n / d) + 1;
         int32_t mod = n % d;
         // Correct result if the / operator above rounded in the wrong direction.
         return quot + (mod > 0) - (mod && round_down);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.23.h
index dbb03bbccc..eb87c9c3b6 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.23.h
@@ -201,7 +201,7 @@ struct FeeFrac
     {
         Assume(size > 0);
         Assume(at_size >= 0);
-        if (fee >= 0 && fee < 0x200000000) [[likely]] {
+                if (fee >= 0 && fee > 0x200000000) [[likely]] {
             // Common case where (this->fee * at_size) is guaranteed to fit in a uint64_t.
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.0.h
index dbb03bbccc..41599bcccd 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.0.h
@@ -216,7 +216,7 @@ struct FeeFrac
 
 public:
     /** Compute the fee for a given size `at_size` using this object's feerate, rounding down. */
-    int64_t EvaluateDown(int32_t at_size) const noexcept { return Evaluate<true>(at_size); }
+        int64_t EvaluateDown(int32_t at_size) const noexcept { return Evaluate<false>(at_size); }
     /** Compute the fee for a given size `at_size` using this object's feerate, rounding up. */
     int64_t EvaluateUp(int32_t at_size) const noexcept { return Evaluate<false>(at_size); }
 };

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.33.h
index dbb03bbccc..7776bb61da 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.33.h
@@ -69,7 +69,7 @@ struct FeeFrac
         // that the / operator here rounds in the wrong direction (if n_low is not a multiple of
         // size, and is (if round_down) negative, or (if !round_down) positive). If so, make a
         // correction.
-        int64_t quot_low = n_low / d;
+                int64_t quot_low = (n_low / d) - 1;
         int32_t mod_low = n_low % d;
         quot_low += (mod_low > 0) - (mod_low && round_down);
         // Combine and return the result

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.16.h
index dbb03bbccc..ec3adc87bf 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.16.h
@@ -71,7 +71,7 @@ struct FeeFrac
         // correction.
         int64_t quot_low = n_low / d;
         int32_t mod_low = n_low % d;
-        quot_low += (mod_low > 0) - (mod_low && round_down);
+                quot_low += (mod_low > 0) - (mod_low || round_down);
         // Combine and return the result
         return (quot_high << 32) + quot_low;
     }

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.12.h
index dbb03bbccc..92848b64d3 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.12.h
@@ -201,7 +201,7 @@ struct FeeFrac
     {
         Assume(size > 0);
         Assume(at_size >= 0);
-        if (fee >= 0 && fee < 0x200000000) [[likely]] {
+                if (fee <= 0 && fee < 0x200000000) [[likely]] {
             // Common case where (this->fee * at_size) is guaranteed to fit in a uint64_t.
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.18.h
index dbb03bbccc..08f5b1fd81 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.18.h
@@ -201,7 +201,7 @@ struct FeeFrac
     {
         Assume(size > 0);
         Assume(at_size >= 0);
-        if (fee >= 0 && fee < 0x200000000) [[likely]] {
+                if (fee >= 0 || fee < 0x200000000) [[likely]] {
             // Common case where (this->fee * at_size) is guaranteed to fit in a uint64_t.
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.4.h
index dbb03bbccc..2177b00401 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.4.h
@@ -92,7 +92,7 @@ struct FeeFrac
     {
         Assume(d > 0);
         // Compute the division.
-        int64_t quot = n / d;
+                int64_t quot = n * d;
         int32_t mod = n % d;
         // Correct result if the / operator above rounded in the wrong direction.
         return quot + (mod > 0) - (mod && round_down);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.24.h
index dbb03bbccc..ac4672aa71 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.24.h
@@ -201,7 +201,7 @@ struct FeeFrac
     {
         Assume(size > 0);
         Assume(at_size >= 0);
-        if (fee >= 0 && fee < 0x200000000) [[likely]] {
+                if (fee >= 0 && fee >= 0x200000000) [[likely]] {
             // Common case where (this->fee * at_size) is guaranteed to fit in a uint64_t.
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.20.h
index dbb03bbccc..abc34f1f78 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.20.h
@@ -95,7 +95,7 @@ struct FeeFrac
         int64_t quot = n / d;
         int32_t mod = n % d;
         // Correct result if the / operator above rounded in the wrong direction.
-        return quot + (mod > 0) - (mod && round_down);
+                return quot + (mod > 0) + (mod && round_down);
     }
 #else
     static constexpr auto Mul = MulFallback;

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.15.h
index dbb03bbccc..ce3a513935 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.15.h
@@ -95,7 +95,7 @@ struct FeeFrac
         int64_t quot = n / d;
         int32_t mod = n % d;
         // Correct result if the / operator above rounded in the wrong direction.
-        return quot + (mod > 0) - (mod && round_down);
+                return quot + (mod <= 0) - (mod && round_down);
     }
 #else
     static constexpr auto Mul = MulFallback;

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.9.h
index dbb03bbccc..b3563eb3a6 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.9.h
@@ -71,7 +71,7 @@ struct FeeFrac
         // correction.
         int64_t quot_low = n_low / d;
         int32_t mod_low = n_low % d;
-        quot_low += (mod_low > 0) - (mod_low && round_down);
+                quot_low += (mod_low >= 0) - (mod_low && round_down);
         // Combine and return the result
         return (quot_high << 32) + quot_low;
     }

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.3.h
index dbb03bbccc..001523cad9 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.3.h
@@ -69,7 +69,7 @@ struct FeeFrac
         // that the / operator here rounds in the wrong direction (if n_low is not a multiple of
         // size, and is (if round_down) negative, or (if !round_down) positive). If so, make a
         // correction.
-        int64_t quot_low = n_low / d;
+                int64_t quot_low = n_low * d;
         int32_t mod_low = n_low % d;
         quot_low += (mod_low > 0) - (mod_low && round_down);
         // Combine and return the result

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.30.h
index dbb03bbccc..814f0fc845 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.30.h
@@ -93,7 +93,7 @@ struct FeeFrac
         Assume(d > 0);
         // Compute the division.
         int64_t quot = n / d;
-        int32_t mod = n % d;
+                int32_t mod = (n % d) + 1;
         // Correct result if the / operator above rounded in the wrong direction.
         return quot + (mod > 0) - (mod && round_down);
     }

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.7.h
index dbb03bbccc..13bbadc2e3 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.7.h
@@ -71,7 +71,7 @@ struct FeeFrac
         // correction.
         int64_t quot_low = n_low / d;
         int32_t mod_low = n_low % d;
-        quot_low += (mod_low > 0) - (mod_low && round_down);
+                quot_low += (mod_low < 0) - (mod_low && round_down);
         // Combine and return the result
         return (quot_high << 32) + quot_low;
     }

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.34.h
index dbb03bbccc..00f6e048f6 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.34.h
@@ -70,7 +70,7 @@ struct FeeFrac
         // size, and is (if round_down) negative, or (if !round_down) positive). If so, make a
         // correction.
         int64_t quot_low = n_low / d;
-        int32_t mod_low = n_low % d;
+                int32_t mod_low = (n_low % d) - 1;
         quot_low += (mod_low > 0) - (mod_low && round_down);
         // Combine and return the result
         return (quot_high << 32) + quot_low;

```
11.h
```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.11.h
index dbb03bbccc..eb87c9c3b6 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.11.h
@@ -201,7 +201,7 @@ struct FeeFrac
     {
         Assume(size > 0);
         Assume(at_size >= 0);
-        if (fee >= 0 && fee < 0x200000000) [[likely]] {
+                if (fee >= 0 && fee > 0x200000000) [[likely]] {
             // Common case where (this->fee * at_size) is guaranteed to fit in a uint64_t.
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.25.h
index dbb03bbccc..0c3ddaee86 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.25.h
@@ -61,7 +61,7 @@ struct FeeFrac
         // Compute quot_high = n.first / d, so the result becomes
         // (n.second + (n.first - quot_high * d) * 2**32) / d + (quot_high * 2**32), or
         // (n.second + (n.first % d) * 2**32) / d + (quot_high * 2**32).
-        int64_t quot_high = n.first / d;
+                int64_t quot_high = (n.first / d) + 1;
         // Evaluate the parenthesized expression above, so the result becomes
         // n_low / d + (quot_high * 2**32)
         int64_t n_low = ((n.first % d) << 32) + n.second;

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.21.h
index dbb03bbccc..657f9144d4 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.21.h
@@ -206,7 +206,7 @@ struct FeeFrac
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);
             } else {
-                return (uint64_t(fee) * at_size + size - 1U) / uint32_t(size);
+                                return (uint64_t(fee) * at_size + size + 1U) / uint32_t(size);
             }
         } else {
             // Otherwise, use Mul and Div.

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.14.h
index dbb03bbccc..e5cdb1f96f 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.14.h
@@ -71,7 +71,7 @@ struct FeeFrac
         // correction.
         int64_t quot_low = n_low / d;
         int32_t mod_low = n_low % d;
-        quot_low += (mod_low > 0) - (mod_low && round_down);
+                quot_low += (mod_low <= 0) - (mod_low && round_down);
         // Combine and return the result
         return (quot_high << 32) + quot_low;
     }

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.31.h
index dbb03bbccc..fcddd25bb4 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.31.h
@@ -61,7 +61,7 @@ struct FeeFrac
         // Compute quot_high = n.first / d, so the result becomes
         // (n.second + (n.first - quot_high * d) * 2**32) / d + (quot_high * 2**32), or
         // (n.second + (n.first % d) * 2**32) / d + (quot_high * 2**32).
-        int64_t quot_high = n.first / d;
+                int64_t quot_high = (n.first / d) - 1;
         // Evaluate the parenthesized expression above, so the result becomes
         // n_low / d + (quot_high * 2**32)
         int64_t n_low = ((n.first % d) << 32) + n.second;

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.2.h
index dbb03bbccc..071fc90a19 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.2.h
@@ -61,7 +61,7 @@ struct FeeFrac
         // Compute quot_high = n.first / d, so the result becomes
         // (n.second + (n.first - quot_high * d) * 2**32) / d + (quot_high * 2**32), or
         // (n.second + (n.first % d) * 2**32) / d + (quot_high * 2**32).
-        int64_t quot_high = n.first / d;
+                int64_t quot_high = n.first * d;
         // Evaluate the parenthesized expression above, so the result becomes
         // n_low / d + (quot_high * 2**32)
         int64_t n_low = ((n.first % d) << 32) + n.second;

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.8.h
index dbb03bbccc..084326c37d 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.8.h
@@ -95,7 +95,7 @@ struct FeeFrac
         int64_t quot = n / d;
         int32_t mod = n % d;
         // Correct result if the / operator above rounded in the wrong direction.
-        return quot + (mod > 0) - (mod && round_down);
+                return quot + (mod < 0) - (mod && round_down);
     }
 #else
     static constexpr auto Mul = MulFallback;

```
35.h
```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.35.h
index dbb03bbccc..ae18b13a15 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.35.h
@@ -92,7 +92,7 @@ struct FeeFrac
     {
         Assume(d > 0);
         // Compute the division.
-        int64_t quot = n / d;
+                int64_t quot = (n / d) - 1;
         int32_t mod = n % d;
         // Correct result if the / operator above rounded in the wrong direction.
         return quot + (mod > 0) - (mod && round_down);

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.6.h
index dbb03bbccc..6c98522329 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.6.h
@@ -206,7 +206,7 @@ struct FeeFrac
             if constexpr (RoundDown) {
                 return (uint64_t(fee) * at_size) / uint32_t(size);
             } else {
-                return (uint64_t(fee) * at_size + size - 1U) / uint32_t(size);
+                                return (uint64_t(fee) * at_size + size - 1U) * uint32_t(size);
             }
         } else {
             // Otherwise, use Mul and Div.

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.10.h
index dbb03bbccc..92b2c76d10 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.10.h
@@ -95,7 +95,7 @@ struct FeeFrac
         int64_t quot = n / d;
         int32_t mod = n % d;
         // Correct result if the / operator above rounded in the wrong direction.
-        return quot + (mod > 0) - (mod && round_down);
+                return quot + (mod >= 0) - (mod && round_down);
     }
 #else
     static constexpr auto Mul = MulFallback;

```
26.h
```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.26.h
index dbb03bbccc..6b8d17594c 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.26.h
@@ -64,7 +64,7 @@ struct FeeFrac
         int64_t quot_high = n.first / d;
         // Evaluate the parenthesized expression above, so the result becomes
         // n_low / d + (quot_high * 2**32)
-        int64_t n_low = ((n.first % d) << 32) + n.second;
+                int64_t n_low = (((n.first % d) << 32) + n.second) + 1;
         // Evaluate the division so the result becomes quot_low + quot_high * 2**32. It is possible
         // that the / operator here rounds in the wrong direction (if n_low is not a multiple of
         // size, and is (if round_down) negative, or (if !round_down) positive). If so, make a

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.28.h
index dbb03bbccc..e1b7b34e80 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.28.h
@@ -70,7 +70,7 @@ struct FeeFrac
         // size, and is (if round_down) negative, or (if !round_down) positive). If so, make a
         // correction.
         int64_t quot_low = n_low / d;
-        int32_t mod_low = n_low % d;
+                int32_t mod_low = (n_low % d) + 1;
         quot_low += (mod_low > 0) - (mod_low && round_down);
         // Combine and return the result
         return (quot_high << 32) + quot_low;

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.32.h
index dbb03bbccc..384b3f294d 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.32.h
@@ -64,7 +64,7 @@ struct FeeFrac
         int64_t quot_high = n.first / d;
         // Evaluate the parenthesized expression above, so the result becomes
         // n_low / d + (quot_high * 2**32)
-        int64_t n_low = ((n.first % d) << 32) + n.second;
+                int64_t n_low = (((n.first % d) << 32) + n.second) - 1;
         // Evaluate the division so the result becomes quot_low + quot_high * 2**32. It is possible
         // that the / operator here rounds in the wrong direction (if n_low is not a multiple of
         // size, and is (if round_down) negative, or (if !round_down) positive). If so, make a

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.1.h
index dbb03bbccc..f2991cb61f 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.1.h
@@ -218,7 +218,7 @@ public:
     /** Compute the fee for a given size `at_size` using this object's feerate, rounding down. */
     int64_t EvaluateDown(int32_t at_size) const noexcept { return Evaluate<true>(at_size); }
     /** Compute the fee for a given size `at_size` using this object's feerate, rounding up. */
-    int64_t EvaluateUp(int32_t at_size) const noexcept { return Evaluate<false>(at_size); }
+        int64_t EvaluateUp(int32_t at_size) const noexcept { return Evaluate<true>(at_size); }
 };
 
 /** Compare the feerate diagrams implied by the provided sorted chunks data.

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.17.h
index dbb03bbccc..b4f0590650 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.17.h
@@ -95,7 +95,7 @@ struct FeeFrac
         int64_t quot = n / d;
         int32_t mod = n % d;
         // Correct result if the / operator above rounded in the wrong direction.
-        return quot + (mod > 0) - (mod && round_down);
+                return quot + (mod > 0) - (mod || round_down);
     }
 #else
     static constexpr auto Mul = MulFallback;

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.19.h
index dbb03bbccc..c421f87da4 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.19.h
@@ -71,7 +71,7 @@ struct FeeFrac
         // correction.
         int64_t quot_low = n_low / d;
         int32_t mod_low = n_low % d;
-        quot_low += (mod_low > 0) - (mod_low && round_down);
+                quot_low += (mod_low > 0) + (mod_low && round_down);
         // Combine and return the result
         return (quot_high << 32) + quot_low;
     }

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.36.h
index dbb03bbccc..3a52d4728e 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.36.h
@@ -93,7 +93,7 @@ struct FeeFrac
         Assume(d > 0);
         // Compute the division.
         int64_t quot = n / d;
-        int32_t mod = n % d;
+                int32_t mod = (n % d) - 1;
         // Correct result if the / operator above rounded in the wrong direction.
         return quot + (mod > 0) - (mod && round_down);
     }

```

```diff
diff --git a/src/util/feefrac.h b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.5.h
index dbb03bbccc..587cfedecd 100644
--- a/src/util/feefrac.h
+++ b/Users/brunogarcia/projects/bitcoin-core-dev/muts-pr-30535-feefrac-h/feefrac.mutant.5.h
@@ -204,7 +204,7 @@ struct FeeFrac
         if (fee >= 0 && fee < 0x200000000) [[likely]] {
             // Common case where (this->fee * at_size) is guaranteed to fit in a uint64_t.
             if constexpr (RoundDown) {
-                return (uint64_t(fee) * at_size) / uint32_t(size);
+                                return (uint64_t(fee) * at_size) * uint32_t(size);
             } else {
                 return (uint64_t(fee) * at_size + size - 1U) / uint32_t(size);
             }

```

</details>

ismaelsadeeq reviewed (approved):
Code review ACK 9590cdd1afc78ff73fdccd2d016075bf40b6b0e2



> @ismaelsadeeq It's a possibility, if we'd want to replace CFeeRate entirely. Another possibility is keeping CFeeRate and its interface, but make it be an encapsulated FeeFrac object (that e.g. on serialization converts to sats/kvb, but that otherwise keeps exact fractions).

I prefer the idea of encapsulating `FeeFrac` within `CFeeRate` given how widely `CFeeRate` is relied upon. I’ve attempt to implement this approach on top of this PR.
If this aligns with your idea, I can create a PR for it although the test currently assumes `CFeeRate` represents a fee rate per kb rather than a fraction, so I had to touch the test a bit.

Note: A few mempool tests are failing with this diff, which I haven’t yet investigated.

<details>
<summary>see diff</summary>

```diff
diff --git a/src/policy/feerate.cpp b/src/policy/feerate.cpp
index eb0cba5c67a..05b4905b5e3 100644
--- a/src/policy/feerate.cpp
+++ b/src/policy/feerate.cpp
@@ -9,37 +9,24 @@
 
 #include <cmath>
 
-CFeeRate::CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes)
+CAmount CFeeRate::GetFee(uint32_t num_bytes) const
 {
-    const int64_t nSize{num_bytes};
-
-    if (nSize > 0) {
-        nSatoshisPerK = nFeePaid * 1000 / nSize;
-    } else {
-        nSatoshisPerK = 0;
-    }
+    return feerate.EvaluateUp(num_bytes);
 }
 
-CAmount CFeeRate::GetFee(uint32_t num_bytes) const
+CAmount CFeeRate::GetFeePerK() const
 {
-    const int64_t nSize{num_bytes};
-
-    // Be explicit that we're converting from a double to int64_t (CAmount) here.
-    // We've previously had issues with the silent double->int64_t conversion.
-    CAmount nFee{static_cast<CAmount>(std::ceil(nSatoshisPerK * nSize / 1000.0))};
-
-    if (nFee == 0 && nSize != 0) {
-        if (nSatoshisPerK > 0) nFee = CAmount(1);
-        if (nSatoshisPerK < 0) nFee = CAmount(-1);
+    if (feerate.size > 0) {
+        return feerate.fee * kvb / feerate.size;
     }
-
-    return nFee;
+    return 0;
 }
 
 std::string CFeeRate::ToString(const FeeEstimateMode& fee_estimate_mode) const
 {
+    CAmount nSatoshisPerK = GetFeePerK();
     switch (fee_estimate_mode) {
-    case FeeEstimateMode::SAT_VB: return strprintf("%d.%03d %s/vB", nSatoshisPerK / 1000, nSatoshisPerK % 1000, CURRENCY_ATOM);
+    case FeeEstimateMode::SAT_VB: return strprintf("%d.%03d %s/vB", nSatoshisPerK / kvb, nSatoshisPerK % kvb, CURRENCY_ATOM);
     default:                      return strprintf("%d.%08d %s/kvB", nSatoshisPerK / COIN, nSatoshisPerK % COIN, CURRENCY_UNIT);
     }
 }
diff --git a/src/policy/feerate.h b/src/policy/feerate.h
index 2e501729148..16c76fa71e5 100644
--- a/src/policy/feerate.h
+++ b/src/policy/feerate.h
@@ -8,6 +8,7 @@
 
 #include <consensus/amount.h>
 #include <serialize.h>
+#include <util/feefrac.h>
 
 
 #include <cstdint>
@@ -26,20 +27,21 @@ enum class FeeEstimateMode {
     SAT_VB,       //!< Use sat/vB fee rate unit
 };
 
+const int32_t kvb{1000};
 /**
  * Fee rate in satoshis per kilovirtualbyte: CAmount / kvB
  */
 class CFeeRate
 {
 private:
-    /** Fee rate in sat/kvB (satoshis per 1000 virtualbytes) */
-    CAmount nSatoshisPerK;
+    FeeFrac feerate;
 
 public:
     /** Fee rate of 0 satoshis per kvB */
-    CFeeRate() : nSatoshisPerK(0) { }
+    CFeeRate() : feerate(0, kvb) { }
+    /** Fee rate of _nSatoshisPerK satoshis per kvB */
     template<typename I>
-    explicit CFeeRate(const I _nSatoshisPerK): nSatoshisPerK(_nSatoshisPerK) {
+    explicit CFeeRate(const I _nSatoshisPerK): feerate(_nSatoshisPerK, kvb) {
         // We've previously had bugs creep in from silent double->int conversion...
         static_assert(std::is_integral<I>::value, "CFeeRate should be used without floats");
     }
@@ -50,7 +52,7 @@ public:
      * param@[in]   nFeePaid    The fee paid by a transaction, in satoshis
      * param@[in]   num_bytes   The vsize of a transaction, in vbytes
      */
-    CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes);
+    CFeeRate(const CAmount& nFeePaid, uint32_t num_bytes) : feerate(nFeePaid, num_bytes) {};
 
     /**
      * Return the fee in satoshis for the given vsize in vbytes.
@@ -62,19 +64,19 @@ public:
     /**
      * Return the fee in satoshis for a vsize of 1000 vbytes
      */
-    CAmount GetFeePerK() const { return nSatoshisPerK; }
-    friend bool operator<(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK < b.nSatoshisPerK; }
-    friend bool operator>(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK > b.nSatoshisPerK; }
-    friend bool operator==(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK == b.nSatoshisPerK; }
-    friend bool operator<=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK <= b.nSatoshisPerK; }
-    friend bool operator>=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK >= b.nSatoshisPerK; }
-    friend bool operator!=(const CFeeRate& a, const CFeeRate& b) { return a.nSatoshisPerK != b.nSatoshisPerK; }
-    CFeeRate& operator+=(const CFeeRate& a) { nSatoshisPerK += a.nSatoshisPerK; return *this; }
+    CAmount GetFeePerK() const;
+    friend bool operator<(const CFeeRate& a, const CFeeRate& b) { return a.feerate < b.feerate; }
+    friend bool operator>(const CFeeRate& a, const CFeeRate& b) { return a.feerate > b.feerate; }
+    friend bool operator==(const CFeeRate& a, const CFeeRate& b) { return a.feerate == b.feerate; }
+    friend bool operator<=(const CFeeRate& a, const CFeeRate& b) { return a.feerate <= b.feerate; }
+    friend bool operator>=(const CFeeRate& a, const CFeeRate& b) { return a.feerate >= b.feerate; }
+    friend bool operator!=(const CFeeRate& a, const CFeeRate& b) { return a.feerate != b.feerate; }
+    CFeeRate& operator+=(const CFeeRate& a) { feerate += a.feerate; return *this; }
     std::string ToString(const FeeEstimateMode& fee_estimate_mode = FeeEstimateMode::BTC_KVB) const;
-    friend CFeeRate operator*(const CFeeRate& f, int a) { return CFeeRate(a * f.nSatoshisPerK); }
-    friend CFeeRate operator*(int a, const CFeeRate& f) { return CFeeRate(a * f.nSatoshisPerK); }
+    friend CFeeRate operator*(const CFeeRate& f, int a) { return CFeeRate(a * f.GetFeePerK()); }
+    friend CFeeRate operator*(int a, const CFeeRate& f) { return CFeeRate(a * f.GetFeePerK()); }
 
-    SERIALIZE_METHODS(CFeeRate, obj) { READWRITE(obj.nSatoshisPerK); }
+    SERIALIZE_METHODS(CFeeRate, obj) { READWRITE(obj.GetFeePerK()); }
 };
 
 #endif // BITCOIN_POLICY_FEERATE_H
diff --git a/src/test/amount_tests.cpp b/src/test/amount_tests.cpp
index e5ab1cfb902..ce8aa7f7515 100644
--- a/src/test/amount_tests.cpp
+++ b/src/test/amount_tests.cpp
@@ -50,7 +50,7 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
     feeRate = CFeeRate(123);
     // Rounds up the result, if not integer
     BOOST_CHECK_EQUAL(feeRate.GetFee(0), CAmount(0));
-    BOOST_CHECK_EQUAL(feeRate.GetFee(8), CAmount(1)); // Special case: returns 1 instead of 0
+    BOOST_CHECK_EQUAL(feeRate.GetFee(8), CAmount(1)); // returns 1 instead of 0; because we are rounding up
     BOOST_CHECK_EQUAL(feeRate.GetFee(9), CAmount(2));
     BOOST_CHECK_EQUAL(feeRate.GetFee(121), CAmount(15));
     BOOST_CHECK_EQUAL(feeRate.GetFee(122), CAmount(16));
@@ -61,7 +61,7 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
     feeRate = CFeeRate(-123);
     // Truncates the result, if not integer
     BOOST_CHECK_EQUAL(feeRate.GetFee(0), CAmount(0));
-    BOOST_CHECK_EQUAL(feeRate.GetFee(8), CAmount(-1)); // Special case: returns -1 instead of 0
+    BOOST_CHECK_EQUAL(feeRate.GetFee(8), CAmount(0));
     BOOST_CHECK_EQUAL(feeRate.GetFee(9), CAmount(-1));
 
     // check alternate constructor
@@ -70,21 +70,21 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
     BOOST_CHECK_EQUAL(feeRate.GetFee(100), altFeeRate.GetFee(100));
 
     // Check full constructor
-    BOOST_CHECK(CFeeRate(CAmount(-1), 0) == CFeeRate(0));
-    BOOST_CHECK(CFeeRate(CAmount(0), 0) == CFeeRate(0));
-    BOOST_CHECK(CFeeRate(CAmount(1), 0) == CFeeRate(0));
+    BOOST_CHECK(CFeeRate(CAmount(-1)).GetFeePerK() == -1);
+    BOOST_CHECK(CFeeRate(CAmount(0)).GetFeePerK() == 0);
+    BOOST_CHECK(CFeeRate(CAmount(1), kvb).GetFeePerK() == 1);
     // default value
     BOOST_CHECK(CFeeRate(CAmount(-1), 1000) == CFeeRate(-1));
     BOOST_CHECK(CFeeRate(CAmount(0), 1000) == CFeeRate(0));
     BOOST_CHECK(CFeeRate(CAmount(1), 1000) == CFeeRate(1));
     // lost precision (can only resolve satoshis per kB)
-    BOOST_CHECK(CFeeRate(CAmount(1), 1001) == CFeeRate(0));
-    BOOST_CHECK(CFeeRate(CAmount(2), 1001) == CFeeRate(1));
+    BOOST_CHECK(CFeeRate(CAmount(1), 1001).GetFeePerK() == 0);
+    BOOST_CHECK(CFeeRate(CAmount(2), 1001).GetFeePerK() == 1);
     // some more integer checks
-    BOOST_CHECK(CFeeRate(CAmount(26), 789) == CFeeRate(32));
-    BOOST_CHECK(CFeeRate(CAmount(27), 789) == CFeeRate(34));
+    BOOST_CHECK(CFeeRate(CAmount(26), 789).GetFeePerK() == CFeeRate(32).GetFeePerK());
+    BOOST_CHECK(CFeeRate(CAmount(27), 789).GetFeePerK() == CFeeRate(34).GetFeePerK());
     // Maximum size in bytes, should not crash
-    CFeeRate(MAX_MONEY, std::numeric_limits<uint32_t>::max()).GetFeePerK();
+    CFeeRate(MAX_MONEY, std::numeric_limits<int32_t>::max()).GetFeePerK();
 
     // check multiplication operator
     // check multiplying by zero
@@ -105,8 +105,8 @@ BOOST_AUTO_TEST_CASE(GetFeeTest)
     // check boundary values
     int maxInt = std::numeric_limits<int>::max();
     feeRate = CFeeRate(maxInt);
-    BOOST_CHECK(feeRate * 2 == CFeeRate(static_cast<int64_t>(maxInt) * 2));
-    BOOST_CHECK(2 * feeRate == CFeeRate(static_cast<int64_t>(maxInt) * 2));
+    BOOST_CHECK((feeRate * 2).GetFeePerK() == CFeeRate(static_cast<int64_t>(maxInt) * 2).GetFeePerK());
+    BOOST_CHECK((2 * feeRate).GetFeePerK() == CFeeRate(static_cast<int64_t>(maxInt) * 2).GetFeePerK());
     // check with zero fee rate
     feeRate = CFeeRate(0);
     BOOST_CHECK(feeRate * 5 == CFeeRate(0));
@@ -125,9 +125,9 @@ BOOST_AUTO_TEST_CASE(BinaryOperatorTest)
     BOOST_CHECK(a <= a);
     BOOST_CHECK(b >= a);
     BOOST_CHECK(b >= b);
-    // a should be 0.00000002 BTC/kvB now
+    // a is still 0.00000001 BTC/kvB because the size is added also
     a += a;
-    BOOST_CHECK(a == b);
+    BOOST_CHECK_EQUAL(a.GetFeePerK(), a.GetFeePerK());
 }
 
 BOOST_AUTO_TEST_CASE(ToStringTest)

```
</details>

glozow commented on src/util/feefrac.h:
```diff
@@ -186,23 +190,35 @@ struct FeeFrac
      * Requires this->size > 0, at_size >= 0, and that the correct result fits in a int64_t. This
      * is guaranteed to be the case when 0 <= at_size <= this->size.
      */
+    template<bool RoundDown>
```
Question: why did you make this a templated function instead of adding a bool parameter? Is it a compile-time optimization?

glozow replied:
Makes sense!

fanquake commented on src/util/feefrac.h:
```diff
@@ -144,6 +186,39 @@ struct FeeFrac
+    {
+        Assume(size > 0);
+        Assume(at_size >= 0);
+        if (fee >= 0 && fee < 0x200000000) [[likely]] {
```
Was curious how much difference using `[[likely]]` may/may-not have, as I've seen mixed discussions around it's usage. For a single call to `EvaluateDown` (from the profile_estimate pass):

![combined](https://github.com/user-attachments/assets/83eb899c-e4b2-4bff-a402-f4a45109a3f2)

paplorinc replied:
I personally view it as documentation and would definitely keep it here

paplorinc commented on src/arith_uint256.h:
```diff
@@ -210,13 +211,8 @@ class base_uint
-    friend inline bool operator<=(const base_uint& a, const base_uint& b) { return a.CompareTo(b) <= 0; }
+    friend inline std::strong_ordering operator<=>(const base_uint& a, const base_uint& b) { return a.CompareTo(b) <=> 0; }
     friend inline bool operator==(const base_uint& a, uint64_t b) { return a.EqualTo(b); }
-    friend inline bool operator!=(const base_uint& a, uint64_t b) { return !a.EqualTo(b); }
```
Checked the same with a simple logging to make sure the reproducer is representative:
```diff
diff --git a/src/arith_uint256.h b/src/arith_uint256.h
index d91689c3b3..0198edbbbe 100644
--- a/src/arith_uint256.h
+++ b/src/arith_uint256.h
@@ -12,6 +12,7 @@
 #include <limits>
 #include <stdexcept>
 #include <string>
+#include <iostream>
 
 class uint256;
 
@@ -212,7 +213,10 @@ public:
     friend inline base_uint operator*(const base_uint& a, uint32_t b) { return base_uint(a) *= b; }
     friend inline bool operator==(const base_uint& a, const base_uint& b) { return memcmp(a.pn, b.pn, sizeof(a.pn)) == 0; }
     friend inline std::strong_ordering operator<=>(const base_uint& a, const base_uint& b) { return a.CompareTo(b) <=> 0; }
-    friend inline bool operator==(const base_uint& a, uint64_t b) { return a.EqualTo(b); }
+    friend inline bool operator==(const base_uint& a, uint64_t b) {
+        std::cout << "base_uint::operator==(const base_uint& a, uint64_t b)" << std::endl;
+        return a.EqualTo(b);
+    }
 
     std::string GetHex() const;
     std::string ToString() const;
diff --git a/src/test/feefrac_tests.cpp b/src/test/feefrac_tests.cpp
index 3bbc7450ea..255e96af01 100644
--- a/src/test/feefrac_tests.cpp
+++ b/src/test/feefrac_tests.cpp
@@ -6,9 +6,19 @@
 #include <random.h>
 
 #include <boost/test/unit_test.hpp>
+#include "arith_uint256.h"
+#include <iostream>
 
 BOOST_AUTO_TEST_SUITE(feefrac_tests)
 
+BOOST_AUTO_TEST_CASE(feefrac_operators_ne)
+{
+    auto x = arith_uint256{};
+    std::cout << x.ToString() << std::endl;
+    std::cout << (x == uint64_t(123)) << std::endl;
+    std::cout << (x != uint64_t(123)) << std::endl;
+}
+
 BOOST_AUTO_TEST_CASE(feefrac_operators)
 {
     FeeFrac p1{1000, 100}, p2{500, 300};
```

>  % make -j10 && src/test/test_bitcoin --run_test=feefrac_tests/feefrac_operators_ne

which prints:
> 0000000000000000000000000000000000000000000000000000000000000000
base_uint::operator==(const base_uint& a, uint64_t b)
0
base_uint::operator==(const base_uint& a, uint64_t b)
1

paplorinc commented:
ACK eff5bf7d673c797d63c5ad15ac18b8316dea5ffe
(diff only contains `Evaluate*` -> `EvaluateFee*` changes)

ismaelsadeeq reviewed (commented):
re-ACK eff5bf7d673c797d63c5ad15ac18b8316dea5ffe

Christewart commented:
concept ACK

l0rinc commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -13,43 +14,34 @@
-    for (size_t i = 0; i < a.size(); ++i) {
-        if (a[i] != b[i]) return a[i] <=> b[i];
+/** Construct an arith_uint256 whose value equals abs(x), for 96-bit x. */
+arith_uint256 Abs256(std::pair<int64_t, uint32_t> x)
```
Now that Fuzz is always compiled, this should likely be move to `feefrac: add helper functions for 96-bit division` (see https://github.com/bitcoin/bitcoin/actions/runs/13162889382/job/36735875466?pr=30535#step:6:2772)

l0rinc reviewed (approved):
Rebase didn't result in any code change, but CI compiles fuzzing for every commit now

l0rinc commented:
ACK 5600091e8384a69cfc7e95a7aea816e7a9076917

Build failure will likely be resolved by retriggering it later

l0rinc commented:
ACK f6aa28cf8fad6a3240498b500524bb380855b18e

Added build failure to https://github.com/bitcoin/bitcoin/issues/31797#issuecomment-2637727127 - should be retriggered later

murchandamus reviewed (commented):
Concept ACK, I also really like the idea of improving CFeeRate.

ACK up to 5147b5d602d1f55d7c9f497cf0cfc23263f5caf5, still mulling over the last three commits.

ismaelsadeeq commented:
@murchandamus 

> Concept ACK, I also really like the idea of improving CFeeRate.

You will also like to https://github.com/bitcoin/bitcoin/pull/31363/commits/46ffaf98503c7f6c8a8d9b9a188f541d77e08b1a

> If this aligns with your idea, I can create a PR for it although the test currently assumes `CFeeRate` represents a fee rate per kb rather than a fraction, so I had to touch the test a bit.



I think after this and #31363 making `CFeeRate` a wrapper around `FeePerVSize` will be a good follow-up.

instagibbs commented:
I did some *very* light fuzzing against `CFeeRate` results to make sure the "gap" between the two values is never huge, would there be some value in adding coverage to confirm that results aren't wildly off from what we have today if we use it more widely?

instagibbs commented:
https://github.com/instagibbs/bitcoin/commit/ac2e55980bb4841a55b9396070af62cf3925e721

Bounding difference to "100 sats", seems to work ok to a 500BTC fee, feel free to take or leave it since these numbers are kind of made up.

concept ACK anyways

laanwj reviewed (approved):
Code review ACK f6aa28cf8fad6a3240498b500524bb380855b18e, mul/division code looks correct, testing and fuzzing coverage seems very good

yancyribbens commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -121,3 +104,89 @@ FUZZ_TARGET(feefrac)
+    auto num_abs = Abs256(num);
+    auto den_abs = Abs256(den);
+    auto quot_abs = (is_negative == round_down) ?
+        (num_abs + den_abs - 1) / den_abs :
```
At first glance this is confusing.  It looks like `quot_abs` is being computed differently depending on if `num_high` is negative or not, although it's not apparent why right away.  Maybe I will take these fuzz tests for a spin and look at some example values.  As a side note, it's interesting that fuzz tests are used to test for program correctness where I thought the main objective for fuzz tests was to find crashes and panics.

yancyribbens replied:
> They're an amazing tool for this purpose. Any time you write a well-contained, well-specified piece of code, you can essentially write a less efficient simulator for its behavior, and compare the two in a fuzz test. We have many such examples in the codebase.

However property tests, which like fuzz tests generate random input, although the inputs are truly random, which may be desirable for testing program correctness.  From what I understand about fuzz tests, the random inputs are generated in such a way that new code paths are discovered with the goal of producing an overflow or crash.

l0rinc commented on src/test/feefrac_tests.cpp:
```diff
@@ -17,6 +17,44 @@ BOOST_AUTO_TEST_CASE(feefrac_operators)
     FeeFrac empty{0, 0};
     [[maybe_unused]] FeeFrac zero_fee{0, 1}; // zero-fee allowed
```
This is [always used now](https://github.com/bitcoin/bitcoin/commit/0606e66fdbb914f984433d8950f0c32b5fb871f3#diff-eca436ae09a12f8a59ac888c7f4ee6f36e4d12f38c2dab47fee0377ce69fe660R18), right?

sipa committed:
fuzz: assert min diff between FeeFrac and CFeeRate

Co-Authored-By: Greg Sanders <gsanders87@gmail.com>

l0rinc commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -121,3 +106,128 @@ FUZZ_TARGET(feefrac)
+    assert(Abs256(res) == quot_abs);
+
+    // Compare approximately with floating-point.
+    long double expect = round_down ? std::floorl(num_high * 4294967296.0L + num_low) / den
+                                    : std::ceill(num_high * 4294967296.0L + num_low) / den;
```
It seems to me that `std::floor` and `std::ceil` already have overloads for `long double`, which we could use instead to appease the CI gods:

> feefrac.cpp:144:44: error: ‘floorl’ is not a member of ‘std’; did you mean ‘floor’?

l0rinc commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -121,3 +106,128 @@ FUZZ_TARGET(feefrac)
+    // Compare approximately with floating-point.
+    long double expect = round_down ? std::floor(num_high * 4294967296.0L + num_low) / den
+                                    : std::ceil(num_high * 4294967296.0L + num_low) / den;
+    // Expect to be accurate within 50 bits of precision, +- 1 sat.
```
I was wondering if we could measure [ULP tolerance](https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition) here instead to better control the step sizes - or if that would be too much effort in this context.

instagibbs replied:
I think there's also some value in knowing we're not diverging wildly from what already exists if we're planning on swapping functionality later. Agreed on the rest.

instagibbs replied:
oh right, I conflated the two threads here

l0rinc reviewed (approved):
ACK cf00540f54302e37f8af08f15fce16cbcab0f8c9

Added approximately floating point comparisons to fuzz tests and removed the lingering `[[maybe_unused]]` in `feefrac_tests.cpp`

ryanofsky commented:
PR may be ready for merge if @ismaelsadeeq and @laanwj want to re-ack or others want to ACK

ismaelsadeeq commented:
I will review soon.

ismaelsadeeq commented on src/test/fuzz/feefrac.cpp:
```diff
@@ -217,5 +218,16 @@ FUZZ_TARGET(feefrac_mul_div)
+            quot_abs <= arith_uint256{std::numeric_limits<int64_t>::max() / 1000}) {
+            CFeeRate feerate(mul64, (uint32_t)div);
+            CAmount feerate_fee{feerate.GetFee(mul32)};
+            auto allowed_gap = static_cast<int64_t>(mul32 / 1000 + 3 + round_down);
```
In "fuzz: assert min diff between FeeFrac and CFeeRate" how did you come up with 1000 and 3 here?
I could not comprehend the computation here. how did you come up with this gap.

ismaelsadeeq reviewed (commented):
Code review ACK cf00540f54302e37f8af08f15fce16cbcab0f8c9

I have reviewed each commit it looks good, my comments below are not serious just a question on last commit and some suggestions that improve doc, making fuzz test dry and improving commit message.

l0rinc commented:
ACK 58914ab459c46c518c47c5082aec25ac0d03ab11

Rebased, changing mostly `Span` to `std::span` - and a minor comment update.

<details>
<summary>patch</summary>

```patch
diff --git a/src/util/feefrac.h b/src/util/feefrac.h
--- a/src/util/feefrac.h(revision a9f419ef8aa2d518de1ad87603b82248bf88f864)
+++ b/src/util/feefrac.h(date 1744038722170)
@@ -84,7 +84,8 @@
         return __int128{a} * b;
     }
 
-    /** Helper function for 96/32 signed division, rounding towards negative infinity. This is a
+    /** Helper function for 96/32 signed division, rounding towards negative infinity (if
+     *  round_down), or towards positive infinity (if !round_down). This is a
      *  version relying on __int128.
      *
      * The result must fit in an int64_t, and d must be strictly positive. */
```

</details>

ismaelsadeeq reviewed (commented):
reACK  58914ab459c46c518c47c5082aec25ac0d03ab11

yancyribbens commented on src/util/feefrac.h:
```diff
@@ -144,6 +187,39 @@ struct FeeFrac
+
+    /** Compute the fee for a given size `at_size` using this object's feerate.
+     *
+     * This effectively corresponds to evaluating (this->fee * at_size) / this->size, with the
```
Wouldn't it save some confusion if there was a type for `WU` and a type for vbytes?  If for example a `FeeFrac` is created as {sats /vbytes} then one wouldn't want to accidentally pass an argument calculated using a `wu` metric for `at_size` param (possibly).  In that case one would want to div by 4 first to get corresponding vbytes metric, however if there where strong types, then it could be converted automagically.

glozow reviewed (commented):
light code review ACK 58914ab459c46c518c47c5082aec25ac0d03ab11

rebroad commented on src/util/feefrac.h:
```diff
@@ -186,23 +191,35 @@ struct FeeFrac
      * Requires this->size > 0, at_size >= 0, and that the correct result fits in a int64_t. This
      * is guaranteed to be the case when 0 <= at_size <= this->size.
      */
+    template<bool RoundDown>
```
snake_case?
